Working on #2972 Template Conversion to Twig Format (calendar.php)

PaulBender opened the pull request:
Working on #2972

This PR is for the calendar (calendar.php).

euantorano commented on calendar.php:
```diff
@@ -242,316 +242,229 @@
+        $years[] = $year;
+    }
+
+    $timezones = build_timezone_select("timezone", $timezone);
```
I wonder if this is worth making a Twig function - do we use the same function elsewhere (such as the User CP)?

PaulBender replied:
As far as I'm aware it's the same everywhere. So it could be a Twig function

euantorano replied:
Alright, sounds good. Not sure whether we should have a `TimeDateExtension`  for this and `my_date` (and whether we have any other time/date related functions to add to it) or if we should just create a `CoreExtension`/`MybbExtension` with these kinds of functions...

euantorano commented on calendar.php:
```diff
@@ -242,316 +242,229 @@
+    // Construct option list for years
+    for ($year = my_date('Y'); $year < (my_date('Y') + 5); ++$year) {
+        if ($year == $single_year) {
+            $select['single_year'] = "selected=\"selected\"";
```
Would be nice to do this logic inside the view rather than building HTML in the backend code. Same goes with the other instances of this type of code.

PaulBender replied:
I agree, but I'm not sure how to do that.

euantorano replied:
Two options:

- Do the conditional check in Twig
- Set a variable such as `$select['single_year_selected'] = true` and in twig do something like:
```twig
<option{%if select.single_year_selected %} selected="selected"{%endif}/>
```

Option two is the easiest by far.

euantorano commented on inc/views/base/calendar/addevent.twig:
```diff
@@ -0,0 +1,404 @@
+            <tr>
+                <td class="trow2" style="vertical-align: top;"><strong>{{ lang.event_date }}</strong></td>
+                <td class="trow1">
+                    <script type="text/javascript">
```
A note for @justinsoltesz @Eric-Jackson - would be nice to take this inline JS out. The less inline JS we have, the better in my opinion.

JordanMussi replied:
Would be great if we could have none so that admins can use a CSP without `unsafe-inline`.

Eric-Jackson replied:
Yeah same goes for inline CSS imo, which is happening after the conversion to Twig I believe. Not sure if it would be better to remove JS now, or wait until we're making the changes to the HTML for a responsive theme.

euantorano commented on inc/views/base/calendar/addevent.twig:
```diff
@@ -0,0 +1,404 @@
+                    <dl style="margin-top: 0; margin-bottom: 0; width: 100%;">
+                        <dt><input type="radio" name="type" value="single" {{ select.type_single }} id="type_single" onclick="toggleType();" /> <label for="type_single"><strong>{{ lang.type_single }}</strong></label></dt>
+                        <dd style="margin-top: 4px;" id="single_selects">
+                            <select name="single_day">
```
Perhaps a good idea to try and abstract this into a loop and possibly break it into a function or partial view or something?

PaulBender commented:
I've finished the conversion work on the calendar. Once I fix the above issues (I hope to do them tomorrow) this will be ready to merge.

One thing I want to ask everyone, should we keep the $gobutton function? I've been including it so far as {{ gobutton|raw }} but it seems pointless. All it does is load a template that contains a submit button, something that can be done inline easily.

euantorano commented:
I just moved it inline for the subscriptions template personally. 

> On 13 Jan 2018, at 23:47, Paul Bender <notifications@github.com> wrote:
> 
> I've finished the conversion work on the calendar. Once I fix the above issues (I hope to do them tomorrow) this will be ready to merge.
> 
> One thing I want to ask everyone, should we keep the $gobutton function? I've been including it so far as {{ gobutton|raw }} but it seems pointless. All it does is load a template that contains a submit button, something that can be done inline easily.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

PaulBender commented:
That should be everything. If there are no other issues this can be merged.

euantorano commented on inc/views/base/calendar/addevent.twig:
```diff
@@ -0,0 +1,320 @@
+                            </select>
+                            &nbsp;
+                            <select name="single_month">
+                                <option value="1"{% if select.single_month.1 %} selected="selected"{% endif %}>{{ lang.month_1 }}</option>
```
Not a big deal, but repeated blocks like these where it's easy to forget to update a single number make me anxious. The problem with using a loop here would be the month language variable, right? If so, you an kind of cheat:

```
{% for i in 1..12 %}
    <option value="{{ i }}"{% if select.single_month[i] %} selected="selected"{% endif %}>{{ attribute(lang, 'month_' ~ i) }}</option>
{% endfor %}
```

The trick here is using the `[]` square brackets for the condition, and the [`attribute`](https://twig.symfony.com/doc/2.x/functions/attribute.html) core Twig function to get the language variable (with string concatenation using the `~` operator).

I __believe__ that should work, but I haven't tested it at all.

euantorano commented on inc/views/base/calendar/calendar.twig:
```diff
@@ -0,0 +1,105 @@
+                        {% if calendar.prev_link %}
+                            <a href="{{ calendar.prev_link|raw }}">&laquo; {{ calendar.prev_month_name }} {{ calendar.prev_month_year }}</a>
+                        {% endif %}
+                        {% if calendar.sep == true %}
```
Could simplify this to `{% if calendar.sep %}`, since you already use the short form for `next_link` below (unless this can be a string or some other type or something...).

euantorano commented on inc/views/base/calendar/calendar.twig:
```diff
@@ -0,0 +1,105 @@
+                            <select name="month">
+                                <option value="{{ calendar.month }}">{{ calendar.currentmonth }}</option>
+                                <option value="{{ calendar.month }}">----------</option>
+                                <option value="1">{{ lang.alt_month_1 }}</option>
```
Again, could use the same loop trick I mentioned in the other template file hopefully.

euantorano commented on inc/views/base/calendar/dayview.twig:
```diff
@@ -0,0 +1,84 @@
+                            <select name="month">
+                                <option value="{{ calendar.month }}">{{ calendar.currentmonth }}</option>
+                                <option value="{{ calendar.month }}">----------</option>
+                                <option value="1">{{ lang.alt_month_1 }}</option>
```
Loop trick again 😉

euantorano commented on inc/views/base/calendar/weekview.twig:
```diff
@@ -0,0 +1,111 @@
+                            <select name="month">
+                                <option value="{{ calendar.week_from_month }}">{{ calendar.week_from_one }}</option>
+                                <option value="{{ calendar.week_from_month }}">----------</option>
+                                <option value="1">{{ lang.alt_month_1 }}</option>
```
Loop again 😄

PaulBender commented:
I've converted the most months to use for loops. The ones I didn't I moved to their own file and included it where necessary (so there's only one instance instead of four).

euantorano commented:
Nice one, thanks. Looks much cleaner now!

euantorano commented:
@PaulBender Can you update the template list in #2972 with the templates you've sorted with this?
